### ISSUE TITLE
Improve assert error messages

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#8200]
### Description

With https://github.com/OpenModelica/OpenModelica/commit/1cbaefaede641a04c749459ab8eca5bdc70a04d6 the error messages for asserts changed and the way the integrator works at an assert.

Before:
```
[ERROR][testAssertFMI] [error][FMU status:Error] fmiCompletedIntegratorStep: terminated by an assertion.
assert            | debug   | fmiCompletedIntegratorStep failed with status : Error
stdout            | warning | Integrator attempt to handle a problem with a called assert.
assert            | warning | The following assertion has been violated at time 0.202853
|                 | |       | x < 1.5
assert            | error   | x is too big at time = 0.202853
[ERROR][testAssertFMI] [error][FMU status:Error] fmiCompletedIntegratorStep: terminated by an assertion.
assert            | debug   | fmiCompletedIntegratorStep failed with status : Error
assert            | warning | The following assertion has been violated at time 0.202853
|                 | |       | x < 1.5
assert            | error   | x is too big at time = 0.202853
[ERROR][testAssertFMI] [error][FMU status:Error] fmiCompletedIntegratorStep: terminated by an assertion.
assert            | debug   | fmiCompletedIntegratorStep failed with status : Error
stdout            | info    | model terminate | Simulation terminated by an assert at time: 0.202853
```

Now:
```
[ERROR][testAssertFMI] [error][FMU status:Error] fmiCompletedIntegratorStep: terminated by an assertion.
assert            | debug   | fmiCompletedIntegratorStep failed with status : Error
stdout            | warning | Integrator attempt to handle a problem with a called assert.
assert            | warning | The following assertion has been violated at time 0.202853
|                 | |       | x < 1.5
assert            | info | x is too big at time = 0.202853
assert            | error   | No event found, but assert was triggered. Throwing now!
```

The message "No event found, but assert was triggered. Throwing now!" is not particularly clear to the uninitiated. Specifically:

### Expected Behavior

Maybe we could make those messages a bit more informative for the layman end-user 😃

### Additional Context

New issue from #6195

**Comment by AnHeuermann** [comment]
@casella wrote in https://github.com/OpenModelica/OpenModelica/issues/6195#issuecomment-975656961
> OK. I have to say that "No event found, but assert was triggered. Throwing now!" is not particularly clear to the uninitiated. Specifically:
> 
> * I guess "no event found but assert was triggered" means that a condition was triggered (x < 1.5 in this case) but it was in an assert, not in an if- or when-clause. Do I get it right?
> * What does "Throwing now!" means? Yeah, I understand that you are throwing an exception when this message is printed on stdout or stderr, but from an end-user perspective this message is a bit puzzling.
> 
> Maybe we could make those messages a bit more informative for the layman end-user 😃

I agree, the message is not very good. Improving the text itself is not an issue. Any suggestions?

-----------------

But actually I want to print error messages only if really needed.

That means we have an assert `A` that is  violated and maybe needs to throw an error. But we will continue for now (without a massage). Then assert `B` is violated as well and maybe needs to throw an error, but we continue.

Later we check if we found an event. In that case we can ignore `A` and `B` and don't need to print any errors.
Otherwise we need the error messages from `A` and `B`. The information `no event was found` isn't really relevant to `A` and `B`. Also it is not clear if we need all messages from all violated asserts or just the first one. I guess all are better, but that can lead to very large error messages.

The C runtime is not very well suited to do all this. We don't have much options for error handling (catching/ collecting / re-throwing).
But I could create some error message buffer and write to that and later print that to `LOG_ASSERT`. But there are no guaranties we always completely run through `simulationUpdate()`. Maybe we are dividing by zero and throw an error. Then all assert messages are lost(?). And I would need to change the error handling a tiny bit.

**Comment by casella** [comment]
What is not clear at all to me is: what is the relationship between errors and events? I don't get it, so the whole thing looks quite meaningless to me. Is it because we are talking of assert violations, which depend on relations in the same way as events? Sorry for being a bit dumb, but I really can't get it.

In any case, my understanding is that events and asserts are completely unrelated. 

Events must be handled as it is well known, i.e. identifying where they happen by means of zero-crossing functions, while staying on one branch of the condition, then switching branch and restarting. 

Asserts are simply meant to terminate the simulation if some of the results are invalid. We don't need to find the exact time an assert is triggered, who cares about that?

The point is that we should check the assertions only on accepted steps, i.e. once all the iterations are over and the tolerances are met. This has nothing to do with event handling, as I understand.

It seems to me that the two concepts were partially mixed up, and this turns out to be unnecessarily complicated to handle. Do I miss something?

**Comment by AnHeuermann** [comment]
> The point is that we should check the assertions only on accepted steps, i.e. once all the iterations are over and the tolerances are met. This has nothing to do with event handling, as I understand.

Yeah, I guess that is our issue here. We evaluate asserts togehter with some other stuff that is needed for zero-crossings. And after that we decide if we can accept the step.

I will see if it makes sense to move the assert check to be the last thing before we accept a step.

**Comment by casella** [comment]
If we could make this into 1.19.0, that would be great. Tentatively adding it to the project

**Comment by casella** [comment]
@AnHeuermann, while you're at it, you may also consider #8235 by @thorade.
